src/devices/RGBLedMatrix/RgbLedMatrix.cs

```diff
@@ -0,0 +1,641 @@
+
+using System;
+using System.Interop;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using System.Device.Gpio;
+using System.Device.Gpio.Drivers;
+using System.Runtime.CompilerServices;
+using Iot.Device.Graphics;
+
+namespace Iot.Device.LEDMatrix
+{
+    public class RGBLedMatrix
+    {
+        private GpioController _controller;
+        private Gpio _gpio; // Gpio handling.
+        private PinMapping _mapping; // mapping the pin numbers to the Gpio (R1, G1, B1, OE, Clock, Latch...etc.).
+        private int _width; // number of visual pixel columns (as passed by the caller).
+        private int _fullChainWidth; // number of physical columns in pixel for the whole chain.
+        private int _rows; // number of visual pixel rows.
+        private int _deviceRows; // number of rows per one matrix.
+        private int _isRendering; // true if we already rendering, false otherwise.
+        private bool _safeToDispose; // true if it is safe to dispose the whole object.
+        private bool _swapRequested; // used internally when switching between the forground and background buffers.
+        private byte[] _colorsBuffer; // Buffer to store all color data used for drawing on the matrix.
+        private byte[] _colorsBackBuffer; // Background buffer to store all color data used for drawing on the matrix.
+        private ulong[] _rowSetMasks; // masks used to quickly set the row number to Gpio pins.
+        private ulong [] _colorsMake; // masks used to quickly set the R, G, B values to Gpio pins.
+        private long _duration = (long) (Stopwatch.Frequency * 1800 / 1E9); // time interval in 100 nanoseconds used in the pulse width modulation (PWM)
+        private long _frameTime; // store the time spent to render one frame.
+        private bool _startMeasureFrameTime; // enable to measure the time frame
+        private int _chainRows; // how many matrices in one column of the chaining.
+        private int _chainColumns; // how many matrices in one row of the chaining.
+
+        /// <summary>
+        /// Construct a new RGBLedMatrix object
+        /// </summary>
+        /// <param name="mapping">The Gpio pin mapping</param>
+        /// <param name="width">The width in pixels of the matrix display area</param>
+        /// <param name="height">The height in pixels of the matrix display area</param>
+        /// <param name="chainRows">Number of the matrices rows in the chain</param>
+        /// <param name="chainColumns">Number of the matrices columns in the chain</param>
+        public RGBLedMatrix(PinMapping mapping, int width, int height, int chainRows = 1, int chainColumns = 1)
+        {
+            _chainRows      = chainRows;
+            _chainColumns   = chainColumns;
+
+            _mapping = mapping;
+
+            if ((width | height) < 0 || height < 8 || height % chainRows != 0)
```

Comment:
should we just check `height < 8` and `width < 8` ? (if it's smaller than 8, it's smaller than 0). PS. nice way to check for negative

Reply:
I didn't think if restricting the width to be 8 and up. I don't believe we can have such matrix but nobody knows. if you feel strong about it then I can change it as you suggest.

Reply:
no strong opinion either way